crates/store/src/state.rs

```diff
@@ -653,25 +737,101 @@ impl State {
 
     /// Lists all known nullifiers with their inclusion blocks, intended for testing.
     pub async fn list_nullifiers(&self) -> Result<Vec<(Nullifier, u32)>, DatabaseError> {
-        self.db.select_nullifiers().await
+        self.db.select_all_nullifiers().await
     }
 
     /// Lists all known accounts, with their ids, latest state hash, and block at which the account
     /// was last modified, intended for testing.
     pub async fn list_accounts(&self) -> Result<Vec<AccountInfo>, DatabaseError> {
-        self.db.select_accounts().await
+        self.db.select_all_accounts().await
     }
 
     /// Lists all known notes, intended for testing.
     pub async fn list_notes(&self) -> Result<Vec<NoteRecord>, DatabaseError> {
-        self.db.select_notes().await
+        self.db.select_all_notes().await
     }
 
-    /// Returns details for public (public) account.
+    /// Returns details for public (on-chain) account.
     pub async fn get_account_details(&self, id: AccountId) -> Result<AccountInfo, DatabaseError> {
         self.db.select_account(id).await
     }
 
+    /// Returns account states with details for public accounts.
+    pub async fn get_account_states(
+        &self,
+        requests: &HashMap<AccountId, AccountStateRequest>,
+    ) -> Result<(BlockNumber, Vec<AccountStateResponse>), DatabaseError> {
+        'reload: loop {
+            let (snapshot_block_num, proofs) = {
+                let inner_state = self.inner.read().await;
+                let mut proofs = HashMap::new();
+                for account_id in requests.keys() {
+                    let key = LeafIndex::new_max_depth(*account_id);
+                    let proof = inner_state.account_tree.open(&key);
+
+                    proofs.insert(*account_id, proof);
+                }
+
+                (Self::compute_latest_block_number(&inner_state), proofs)
+            };
+
+            let infos = self.db.select_accounts_by_ids(requests.keys().copied().collect()).await?;
+
+            // TODO: Should we return an error if one or more accounts are not found?
+
+            let mut responses = Vec::with_capacity(requests.len());
+            for info in infos {
+                // Check if the block number in account update from the DB is not newer than the
+                // latest block in in-memory structure we used for getting proofs for accounts.
+                // Otherwise, we need to reload the data and try again.
+                // This situation might happen because we release the lock for in-memory structure
+                // before the loading from DB. We do this to avoid holding the lock for too long.
```

Comment:
I think for now we can just lock the state while we fulfill this request (assuming this simplifies things, of course).

As a longer-term solution, I like the first option - i.e., have `db_select_xxx` at a given block. Performance-wise, it shouldn't be too bad because it is unlikely we'll advance by more than 1 block while fulfilling this request.

Reply:
Thanks guys, but I don't think both solutions give us any benefit at the moment. Current DB scheme is pretty simple and we can query all latest accounts' data by a single request, as we do now. Even if we block the whole state for serving this request, the state could be updated in microseconds after we unlock it and the client will still receive outdated state. This is common problem for blockchains that we can only ensure that state is up-to-date for the given block number, but not absolutely latest for the current blockchain. We should run calls with providing account hashes we know and node must fail if hashes mismatch, so the client will need to update state and repeat (I guess, we already going to do so). Also I don't see any benefits with locking the whole state (stop the world instead of repeat requests in rare cases) and it doesn't seem to lead to significant simplification of the code (we would remove `loop` and that long comment, refactor of accessing proofs and that's it). So I'd prefer to keep the current solution, if you don't mind.

Reply:
> Even if we block the whole state for serving this request, the state could be updated in microseconds after we unlock it and the client will still receive outdated state.

This is actually not a problem. The main thing we want to avoid is the client getting an inconsistent state (i.e., account data does not verify against Merkle paths). Sending a slightly outdated state to clients is fine.

> Once we change our DB scheme to more sophisticated, it would require to make several requests to get the data which we get by single request.

In one of the other comments, I proposed simplifying this request - so, hopefully, we can keep just one DB request for this endpoint.

Reply:
I think with the endpoint simplification we don't even need approach 1. We just get basic account data from the DB in a single request, get the Merkle paths for these accounts, and we are done.

We could keep the loop, but I don't really see a strong need for it (and it is something that we'll likely need to remove in the future).

Separately, we should probably impose some limit on how many accounts can sent with the request. In the transaction kernel, we limit the number of "foreign accounts" for a single transaction to 64. We can probably do the same here.